sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,602 @@
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  // TODO: Add a variant of set_bit for writeonly registers?
+  uint32_t mask = 1 << bit_index;
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_INTR_TEST_REG_OFFSET, mask);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_set_irq_levels(
+    const dif_spi_device_t *spi, uint16_t rx_level, uint16_t tx_level) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t compressed_limit = 0;
+  compressed_limit |= (rx_level & SPI_DEVICE_FIFO_LEVEL_RXLVL_MASK)
+                      << SPI_DEVICE_FIFO_LEVEL_RXLVL_OFFSET;
+  compressed_limit |= (tx_level & SPI_DEVICE_FIFO_LEVEL_TXLVL_MASK)
+                      << SPI_DEVICE_FIFO_LEVEL_TXLVL_OFFSET;
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_FIFO_LEVEL_REG_OFFSET,
+                      compressed_limit);
+
+  return kDifSpiDeviceResultOk;
+}
+
+/**
+ * Parameters for compressing and decompressing a FIFO pointer register.
+ */
+typedef struct fifo_ptr_params {
+  ptrdiff_t offset;
+  ptrdiff_t write_offset;
+  ptrdiff_t read_offset;
+  uint32_t write_mask;
+  uint32_t read_mask;
+} fifo_ptr_params_t;
+
+static const fifo_ptr_params_t kTxFifoParams = {
+    .offset = SPI_DEVICE_TXF_PTR_REG_OFFSET,
+    .write_offset = SPI_DEVICE_TXF_PTR_WPTR_OFFSET,
+    .write_mask = SPI_DEVICE_TXF_PTR_WPTR_MASK,
+    .read_offset = SPI_DEVICE_TXF_PTR_RPTR_OFFSET,
+    .read_mask = SPI_DEVICE_TXF_PTR_RPTR_MASK,
+};
+
+static const fifo_ptr_params_t kRxFifoParams = {
+    .offset = SPI_DEVICE_RXF_PTR_REG_OFFSET,
+    .write_offset = SPI_DEVICE_RXF_PTR_WPTR_OFFSET,
+    .write_mask = SPI_DEVICE_RXF_PTR_WPTR_MASK,
+    .read_offset = SPI_DEVICE_RXF_PTR_RPTR_OFFSET,
+    .read_mask = SPI_DEVICE_RXF_PTR_RPTR_MASK,
+};
+
+/**
+ * An exploded FIFO pointer value, consisting of a |uint11_t|
+ * offset part, and a |uint1_t| phase.
+ */
+typedef struct fifo_ptr {
+  uint16_t offset;
+  bool phase;
+} fifo_ptr_t;
+
+// Masks for extracting the phase bit, and the address part, out of a FIFO
+// offset.
+static const uint16_t kPhaseMask = SPI_DEVICE_BUFFER_SIZE_BYTES;
+static const uint16_t kOffsetMask = SPI_DEVICE_BUFFER_SIZE_BYTES - 1;
+
+static void fifo_ptr_increment(fifo_ptr_t *ptr, uint16_t increment,
+                               uint16_t fifo_len) {
+  uint32_t inc_with_overflow = ptr->offset + increment;
+  // If we would overflow, wrap and flip the overflow bit.
+  if (inc_with_overflow >= fifo_len) {
+    inc_with_overflow -= fifo_len;
+    ptr->phase = !ptr->phase;
+  }
+
+  ptr->offset = inc_with_overflow & kOffsetMask;
+}
+
+/**
+ * A decompressed FIFO pointer register, consisting of a read offset
+ * and a write offset within the FIFO region.
+ *
+ * The offsets themselves are only |uint11_t|, with an additional
+ * 12th "phase" bit used for detecting the wrapparound behavior of
+ * the ring buffer FIFOs.
+ */
+typedef struct fifo_ptrs {
```

Comment:
Please move it together with `fifo_ptr_t` type, it makes sense because they are closely related.

Reply:
They're already together? The only reason fifo_ptr_increment is after `fifo_ptr_t` is because, unlike all the other fifo pointer operations, it operates on a single fifo_ptr_t, and I want to make that clear.

Reply:
In my opinion, it is more readable if you move this type straight after `fifo_ptr_t`, but not a big deal.